### PR TITLE
(PUP-3771) Acceptance remove use of hosts array

### PR DESCRIPTION
**Pull request by johnduarte**
This commit removes the use of the hosts array in tests where it
should not be assumed that all hosts should be tested. In most cases,
it is replaced by the agents array.

### CODE REVIEW

File: acceptance/tests/face/help_test.rb [hunk 1]
```diff
@@ -1,42 +0,0 @@
...
-    next if face == 'help'
-
-    step "Use `puppet help #{face}`"
-
-    on host, puppet("help #{face} | grep USAGE") do
-      assert_match /^USAGE:/, stdout,
-        "There is NO usage section for 'puppet help #{face}'"
-    end
-  end
-end
```
Comment (er0ck) [inline acceptance/tests/face/help_test.rb]: why is this file being removed?  has it been covered by unit tests?  `puppet help` has regressed several times.

File: acceptance/tests/resource/file/source_attribute.rb [hunk 2]
```diff
@@ -1,282 +1,283 @@
...
-    '#{testdir}':;
-    '#{env_dir}':;
-    '#{prod_dir}':;
-    '#{manifest_dir}':;
-    '#{module_dir}':;
-    '#{test_module_dir}':;
-    '#{test_module_manifests_dir}':;
-    '#{test_module_files_dir}':;
-  }
+  step "Setup - create environment and test module"
```
Comment (er0ck) [inline acceptance/tests/resource/file/source_attribute.rb]: i'd like to see closures on `step` as well as `test_name` but it shouldn't hold up this PR

File: acceptance/tests/resource/file/source_attribute.rb [hunk 3]
```diff
@@ -1,282 +1,283 @@
...
+    on agent, puppet( %{apply --modulepath=#{localsource_testdir} #{localsource_test_manifest}} ) do
+      assert_no_match(/content changed/, stdout, "Shouldn't have overwrote any files")
     end
   end
 
-  step "second run should not update any files using apply with puppet:/// URI source"
-  on agent, puppet( %{apply --modulepath=#{localsource_testdir} #{localsource_test_manifest}} ) do
-    assert_no_match(/content changed/, stdout, "Shouldn't have overwrote any files")
-  end
 end
```
Comment (er0ck) [inline acceptance/tests/resource/file/source_attribute.rb]: it's really hard to tell what is going on in this diff.  do you have  TL;DR?

File: acceptance/tests/resource/package/common_package_name_in_different_providers.rb [hunk 4]
```diff
@@ -1,140 +1,140 @@
...
 
-def verify_state(hosts, pkg, state, match)
-  hosts.each do |agent|
-    cmd = rpm_provider(agent)
-    # Note yum lists packages as <name>.<arch>
-    on agent, "#{cmd} list installed" do
-      method(match).call(/^#{pkg}\./, stdout)
-    end
+  def verify_state(hosts, pkg, state, match)
+    hosts.each do |agent|
```
Comment (er0ck) [inline acceptance/tests/resource/package/common_package_name_in_different_providers.rb]: this is confusing.  but i guess it's fine and the tests are still better...

File: acceptance/tests/resource/package/yum.rb [hunk 5]
```diff
@@ -1,186 +1,186 @@
...
+    verify_state(hosts, pkg, '(?!purged|absent)[^\']+', :assert_match)
   end
 
-  step 'Installing a specific version of a known package succeeds' do
-    verify_absent agents, 'guid'
-    apply_manifest_on(agents, 'package {"guid": ensure => "1.0"}') do |result|
-      assert_match('Package[guid]/ensure: created', "#{result.host}: #{result.stdout}")
-    end
+  def verify_absent(hosts, pkg)
+    verify_state(hosts, pkg, '(?:purged|absent)', :assert_no_match)
```
Comment (er0ck) [inline acceptance/tests/resource/package/yum.rb]: looks like these various `verify_*` methods should be abstracted into an acceptance library.

File: acceptance/tests/face/help_test.rb (hunk 1 above)
Comment (johnduarte) [inline acceptance/tests/face/help_test.rb]: Removing as per audit recommendation. The behavior of help for faces is covered here:
https://github.com/puppetlabs/puppet/blob/master/spec/unit/application/face_base_spec.rb#L440-L452

File: acceptance/tests/resource/file/source_attribute.rb (hunk 3 above)
Comment (johnduarte) [inline acceptance/tests/resource/file/source_attribute.rb]: tl;dr this is all indentation for placing the `test_name` in a block.

File: acceptance/tests/face/help_test.rb (hunk 1 above)
Comment (er0ck) [inline acceptance/tests/face/help_test.rb]: which audit recommendation? `audit:low`?
it's regressed before, despite those unit tests, and for good reason (see below).  i humbly disagree with this being low priority.

in addition, those unit tests do nothing more than ensure when you say `-h` or `--help` it calls the help subcommand.  they don't ensure the help subcommand works properly.

i don't disagree these should be unit/integration, but they aren't covered there yet, that i know of.

**Comment by er0ck** [comment]
i'm good with this other than the help test removal.  
we should really ensure `puppet help help help help help` works :egg:

**Comment by johnduarte** [comment]
Hey @er0ck I have restored the help test.

**Pull request merged**